File: mergeSort.cpp
#include <iostream>
#include <vector>

using namespace std;
void merge(std::vector<int> &v,int ls, int le, int rs, int re)
{
	int i=ls; int j=rs; std::vector<int> temp;
	while(i<=le && j<=re)
	{
		if(v[i] <= v[j])
		{
			temp.push_back(v[i]);
			i++;
		}
		else
		{
			temp.push_back(v[j]);
			j++;
		}
	}

	while(i<=le)
	{
		temp.push_back(v[i]);
		i++;
	}
	while(j<=re)
	{
		temp.push_back(v[j]);
		j++;
	}
	for(int k=0; k<temp.size(); k++)
	{
		v[ls+k]=temp[k];
	}
}
void mergesort(std::vector<int> &v,int s,int e)
{
	if (s==e){
		return; 
	}
	mergesort(v, s, (s+e)/2);
	mergesort(v, (s+e)/2+1, e);

	int ls=s;
	int le=(s+e)/2;
	int rs=(s+e)/2+1;
	int re=e;

	merge(v,ls,le,rs,re);

}
int main()
{
	std::vector<int> v;
	v.push_back(2);v.push_back(-1);v.push_back(-4);v.push_back(0);v.push_back(1);v.push_back(1);
	int start=0;
	int end=v.size()-1;

	mergesort(v, start,end);
	for(int i=0;i<v.size();i++)
	{
		cout<<v[i]<<endl;
	}
}

File: mergesort.py
def Merge(arr,p,q,r):
	nl=q-p+1; nr=r-q;

	left=[0]*nl
	right=[0]*nr
	print(p,q,r,nl-1,nr-1)
	for i in range(nl):
		left[i]=arr[p+i]
	for j in range(nr):
		right[j]=arr[q+j+1]
	i=0;j=0;k=p

	while i < nl and j < nr:
		if left[i] < right[j]:
			arr[k] = left[i]
			i += 1¬l¬
			k += 1
		else:
			arr[k] = right[j]
			j += 1
			k += 1
	while i < nl:
		arr[k]=left[i]
		k=k+1
		i=i+1
	while j < nr:
		arr[k]=right[j]
		k=k+1
		j=j+1

def Mergesort(arr,p,r):
	if(p>=r):
		return
	else:
	    q=(p+r)//2;
	print(p,r)
	Mergesort(arr,p,q)
	Mergesort(arr,q+1,r)
	Merge(arr,p,q,r)


arr = [60, 50, 40, 30, 20, 10, 30];
p=0;r=len(arr)-1;
print(len(arr))
sortedarray = Mergesort(arr,p,r);
print("sorted array:", arr)